module/trie.c

```diff
@@ -0,0 +1,291 @@
+{
+    assert(t != NULL);
+
+    unsigned c = (unsigned) current;
+
+    if (t->children[c] == NULL)
+        t->children[c] = new_trie(current);
+
+    return 0;  
+}
+
+/*
+    Inserts word into trie.
+    Parameters:
+     - word: An char array to be inserted into the given trie
+     - t: A pointer to the given trie
+    
+    Returns:
+     - 0 on success, 1 if error occurs.
+    
+    Details:
+     - For each trie, check if entry of the next character exists in the children array:
+          - If so, move into that node in the array
+          - If not, add a new node and move into that node in the array
+     - Then move on to the next character in string
+     - Set the is_word of the last node to 1
+*/
+int insert_string(char *word, struct trie *t)
+{
+    assert(t!=NULL);
+
+    if (*word == '\0'){
+        t->is_word=1;
+        return 0;
+
+    } else {
+
+        char curr = *word;
+
+        int rc = add_node(curr, t);
+        if (rc != 0){
+            fprintf(stderr, "Fail to add_node");
+            return 1;
+        }
+        
+        word++;
+        return insert_string(word, t->children[(unsigned)curr]);
+    }
+}
+
+/* Searches for word in a trie. 
+ *
+ * Returns: 
+ *  - 1 if word is found. 
+ *  - 0 if word is not found at all.
+ *  - -1 if word is found but end node's is_word is 0.
+ */
+int trie_search(char* word, struct trie *t)
+{
+    int len;
+    struct trie* curr;
+    struct trie** next;
+
+    len = strlen(word);
+    curr = t;
+    next = t->children;
+
+    for (int i=0; i<len; i++) {
+        int j = (int) word[i];
+        curr = next[j];
+
+        if (curr == NULL)
+            return NOT_IN_TRIE;
+
+        next = next[j]->children;
+    }
+
+    if (curr->is_word == 1) 
+        return IN_TRIE;
+
+    return PARTIAL_IN_TRIE;
+}
+
+/* ========================= "trie" type commands (Redis wrapper functions) ======================= */
+
+/* TRIE.INSERT key value */
+int TrieInsert_RedisCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+    RedisModule_AutoMemory(ctx); /* Use automatic memory management. */
+    
+    if (argc != 3) return RedisModule_WrongArity(ctx);
+
+	RedisModuleKey *key = RedisModule_OpenKey(ctx,argv[1],
+        REDISMODULE_READ|REDISMODULE_WRITE);
+    int type = RedisModule_KeyType(key);
+	if (type != REDISMODULE_KEYTYPE_EMPTY &&
+        RedisModule_ModuleTypeGetType(key) != trie)
+    {
+        return RedisModule_ReplyWithError(ctx,REDISMODULE_ERRORMSG_WRONGTYPE);
+    }
+    size_t dummy;
+    char *temp = RedisModule_StringPtrLen(argv[2], &dummy);
+    char *empty = "";
+    if (temp == NULL || strcmp(temp, empty) == 0) {
+    	return RedisModule_ReplyWithError(ctx,"ERR invalid value: must be a string");
+    } 
+
+    struct trie *t;
+    /* Create an empty value object if the key is currently empty. */
+    if (type == REDISMODULE_KEYTYPE_EMPTY) {
+    	t = new_trie('\0');
+    	RedisModule_ModuleTypeSetValue(key,trie,t);
```

Comment:
No it's supposed to be trie

Reply:
@DustinoZhang I'm almost positive Valerie is right here. Is this supposed to reference the struct, or an instance of the struct? If it's referencing the struct it needs to be `struct trie`. If this is something else, please explain it here and add a comment because it's not at all clear from reading.